source/slang/slang-ast-base.h

```diff
@@ -23,39 +23,49 @@ class NodeBase : public RefObject
 {
     SLANG_ABSTRACT_CLASS(NodeBase)
 
-        // By default AST types do *not* store the builder. This is called when constructed tho.
-    SLANG_FORCE_INLINE void setASTBuilder(ASTBuilder* astBuilder) { SLANG_UNUSED(astBuilder); }
+        // MUST be called before used. Called automatically via the ASTBuilder.
+        // Note that the astBuilder is not stored in the NodeBase derived types by default.
+    SLANG_FORCE_INLINE void init(ASTNodeType inAstNodeType, ASTBuilder* /* astBuilder*/ ) { astNodeType = inAstNodeType; }
+
+        /// Get the class info 
+    SLANG_FORCE_INLINE const ReflectClassInfo& getClassInfo() const { return *ReflectClassInfo::getInfo(astNodeType); }
 
     SyntaxClass<NodeBase> getClass() { return SyntaxClass<NodeBase>(&getClassInfo()); }
+
+        /// The type of the node. ASTNodeType(-1) is an invalid node type, and shouldn't appear on any
+        /// correctly constructed (through ASTBuilder) NodeBase derived class. 
+        /// The actual type is set when constructed on the ASTBuilder. 
+    ASTNodeType astNodeType = ASTNodeType(-1);
 };
 
 // Casting of NodeBase
 
 template<typename T>
 SLANG_FORCE_INLINE T* dynamicCast(NodeBase* node)
 {
-    return (node && node->getClassInfo().isSubClassOf(T::kReflectClassInfo)) ? static_cast<T*>(node) : nullptr;
+    return (node && ReflectClassInfo::isSubClassOf(node->astNodeType, T::kReflectClassInfo)) ? static_cast<T*>(node) : nullptr;
```

Comment:
Would it make sense to have `astNodeType` be a wrapper type around the `enum` so that it could provide member functions like `isSubClassOf` rather than requiring separate free/static functions?

Reply:
That perhaps using the ReflectClassInfo, in a better way is preferable as it's a class already, and can have suitable methods, allowing the node type to just be an enum.